File: src/train/actionSampling.ts
import * as tf from '@tensorflow/tfjs'

export interface HeadSample {
  /** [B], int32. サンプルされた行動インデックス。 */
  actions: tf.Tensor1D
  /** [B]。サンプルされた行動の、マスク適用後の分布でのlog確率。 */
  logProbs: tf.Tensor1D
  /** [B]。マスク適用後の分布のエントロピー。 */
  entropy: tf.Tensor1D
}

/**
 * `(mask-1)*1e9`の加算バイアス方式でマスクする。mask=1(合法)ならバイアス0、mask=0(非合法)なら
 * バイアス-1e9。float32では`exp(-1e9ish)`は厳密に0.0まで下がる(生の`-Infinity`バイアスだと
 * `0 * -Infinity = NaN`になるが、有限の-1e9なら`0.0 * 有限値 = 0.0`で済む)ため、
 * 非合法な行動の確率・log確率・エントロピーへの寄与がNaNを出さずに正しくゼロになる。
 */
function maskedLogSoftmax(logits: tf.Tensor2D, mask: tf.Tensor2D): tf.Tensor2D {
  const bias = tf.mul(tf.sub(mask, 1), 1e9)
  const masked = tf.add(logits, bias) as tf.Tensor2D
  return tf.logSoftmax(masked)
}

function entropyFromLogProbs(logProbs: tf.Tensor2D): tf.Tensor1D {
  const probs = tf.exp(logProbs)
  return tf.neg(tf.sum(tf.mul(probs, logProbs), -1)) as tf.Tensor1D
}

function gatherLogProb(logProbs: tf.Tensor2D, actions: tf.Tensor1D): tf.Tensor1D {
  const numActions = logProbs.shape[1]
  const oneHot = tf.oneHot(actions, numActions) as tf.Tensor2D
  return tf.sum(tf.mul(logProbs, oneHot), -1) as tf.Tensor1D
}

/** ロールアウト収集時: 新しい行動をサンプルする。move/attackの各ヘッドに独立に使う。 */
export function sampleMaskedCategorical(logits: tf.Tensor2D, mask: tf.Tensor2D, seed?: number): HeadSample {
  return tf.tidy(() => {
    const bias = tf.mul(tf.sub(mask, 1), 1e9)
    const masked = tf.add(logits, bias) as tf.Tensor2D
    const logProbsAll = tf.logSoftmax(masked)
    const sampled = tf.multinomial(masked, 1, seed) as tf.Tensor2D // [B,1]
    const actions = sampled.reshape([logits.shape[0]]) as tf.Tensor1D
    const logProbs = gatherLogProb(logProbsAll, actions)
    const entropy = entropyFromLogProbs(logProbsAll)
    return { actions, logProbs, entropy }
  })
}

/** 決定的な行動選択(評価・対戦向け): マスク適用後のlogitsが最大の行動を選ぶ。サンプリングしない。 */
export function argmaxMaskedCategorical(logits: tf.Tensor2D, mask: tf.Tensor2D): tf.Tensor1D {
  return tf.tidy(() => {
    const bias = tf.mul(tf.sub(mask, 1), 1e9)
    const masked = tf.add(logits, bias) as tf.Tensor2D
    return tf.argMax(masked, -1) as tf.Tensor1D
  })
}

/**
 * PPO更新時: ロールアウトで既に選ばれた行動を、現在(更新中)のポリシーのlogitsで再評価する
 * (サンプルし直さない)。PPO比`exp(newLogProb - oldLogProb)`とエントロピーボーナスに使う。
 */
export function evaluateMaskedCategorical(
  logits: tf.Tensor2D,
  mask: tf.Tensor2D,
  actions: tf.Tensor1D,
): { logProbs: tf.Tensor1D; entropy: tf.Tensor1D } {
  return tf.tidy(() => {
    const logProbsAll = maskedLogSoftmax(logits, mask)
    const logProbs = gatherLogProb(logProbsAll, actions)
    const entropy = entropyFromLogProbs(logProbsAll)
    return { logProbs, entropy }
  })
}

File: training-py/mayhem_rl/action_sampling.py
"""Masked-categorical action sampling/evaluation for the move and attack heads. Mirrors
src/train/actionSampling.ts's masking algebra: an additive `(mask-1)*1e9` logit bias before
softmax (float-safe — avoids the `0 * -inf = nan` trap of a raw `-inf` bias). Cross-language RNG
does *not* need to match between the two sides (see the bridge design's RNG decision) — only the
masking *shape* needs to behave identically, which is why the same bias formula is preserved here
even though PyTorch's own sampling RNG is independent of TF.js's.
"""

from __future__ import annotations

from dataclasses import dataclass

import torch
import torch.nn.functional as F

MASK_BIAS = 1e9


def _masked_logits(logits: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
    bias = (mask - 1.0) * MASK_BIAS
    return logits + bias


def _entropy_from_log_probs(log_probs: torch.Tensor) -> torch.Tensor:
    probs = log_probs.exp()
    return -(probs * log_probs).sum(dim=-1)


def _gather_log_prob(log_probs: torch.Tensor, actions: torch.Tensor) -> torch.Tensor:
    return log_probs.gather(-1, actions.unsqueeze(-1)).squeeze(-1)


@dataclass
class HeadSample:
    actions: torch.Tensor
    log_probs: torch.Tensor
    entropy: torch.Tensor


def sample_masked_categorical(
    logits: torch.Tensor, mask: torch.Tensor, generator: torch.Generator | None = None
) -> HeadSample:
    """Rollout collection: samples a new action. Used independently for the move/attack heads."""
    masked = _masked_logits(logits, mask)
    log_probs_all = F.log_softmax(masked, dim=-1)
    probs_all = log_probs_all.exp()
    actions = torch.multinomial(probs_all, 1, generator=generator).squeeze(-1)
    log_probs = _gather_log_prob(log_probs_all, actions)
    entropy = _entropy_from_log_probs(log_probs_all)
    return HeadSample(actions=actions, log_probs=log_probs, entropy=entropy)


def evaluate_masked_categorical(
    logits: torch.Tensor, mask: torch.Tensor, actions: torch.Tensor
) -> tuple[torch.Tensor, torch.Tensor]:
    """PPO update: re-evaluates an already-chosen action under the current (updated) policy's
    logits, rather than resampling. Used for the PPO ratio `exp(newLogProb - oldLogProb)` and the
    entropy bonus."""
    log_probs_all = F.log_softmax(_masked_logits(logits, mask), dim=-1)
    log_probs = _gather_log_prob(log_probs_all, actions)
    entropy = _entropy_from_log_probs(log_probs_all)
    return log_probs, entropy


def argmax_masked_categorical(logits: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
    """Deterministic action choice (evaluation/replay): the legal action with the highest logit."""
    return torch.argmax(_masked_logits(logits, mask), dim=-1)
